### ISSUE TITLE
`enhanced-img` should not delay JavaScript loading until after all images have finished loading

### ISSUE THREAD

**Issue by OOPS-Studio** [sveltejs/kit#12842]
### Describe the problem

Currently, any `<enhanced:img>` tags that are initialized with the component will have their images loaded before the component's own JavaScript. This means that the page is completely unresponsive until the very last image finishes loading. Combine this with prerendering and the user will be shown a completely-rendered page with buttons and everything - but clicking a button will do nothing because that one image way off the bottom of the screen hasn't finished loading yet.

Delaying the creation of the `<enhanced:img>` tag solves this problem. For example, by placing them inside an `{#if mounted}` block and then `onMount(() => mounted = true);`. Doing this lets the JavaScript load before starting to load the images, but doing so removes the ability to reserve space for the images, which is one of the largest advantages of `enhanced-img`. Since the images aren't prerendered into the HTML, it causes the page to render without them. Then the mount occurs and the images appear and you get a massive content shift.

There needs to be some way to have an `<enhanced:img>` in a component's markup without also making that component's JavaScript wait to load until after the image finishes. The browser-native `<img>` tag does not block the loading of JavaScript. `<enhanced:img>` shouldn't either.

### Describe the proposed solution

Do one of the following:
1. Make `<enhanced:img>` replicate the browser-native `<img>` tag's loading behavior (this is the preferred solution)
2. Make `<enhanced:img>`s not start loading until after all JavaScript finishes
3. Allow both the `<enhanced:img>`s and JavaScript to load at the same time (instead of blocking JavaScript entirely in favor of the images)
4. Allow the developer to choose whatever works best for their page

### Importance

would make my life easier

### Additional Information

I love `enhanced-img`, but this is enough to make me remove it from my project and implement image optimization manually. It is simply unacceptable to have my entire page be unresponsive for 10+ seconds on slow connections. There's no reason a 2kb JavaScript file needs to wait for 1mb of images to load first.

Not only does this make the page unresponsive for large amounts of time, but it also makes trying to apply JavaScript to the images while they load impossible. For example, if I want to display a loading icon over the images while they load and then remove them after each one finishes - this is simply impossible to do. Instead, all the loading icons will remain on all the images (even the ones that have already finished loading) for the entire duration of the page load, only to then all disappear at the same time after the final image loads. It's unusable.

**Comment by benmccann** [comment]
This doesn't sound correct to me. Enhanced image builds HTML picture tags and they're not handled any differently by the browser than hand-crafted HTML.

Can you provide a reproduction and details of what you're looking at to make this determination? Also, is it prod mode or dev mode? And what browser are you using?

**Comment by OOPS-Studio** [comment]
Okay so I just did some more tests and _this only happens in Microsoft Edge._ (I'm running the latest stable version of Edge) It works properly (e.g. allows the JavaScript to load alongside the images) in Opera and Firefox (I do not have Chrome downloaded). I should have included that in my original issue - thank you for reminding me to go check that.

As for your other questions:
This happens in both dev mode and production.
To reproduce, simply use `enhanced-img` exactly as is laid out in [this page of the Svelte docs](https://kit.svelte.dev/docs/images#sveltejs-enhanced-img-setup). In my case I'm specifically using the Vite `import.meta.glob()` method - which may be affecting things. I'm using the static adapter for building, but that probably doesn't matter since this happens in dev mode as well.
As for what I'm looking at, I included screenshots of the Network tab in Chrome DevTools (running inside Edge) showing how every single `.avif` file finishes loading before the JavaScript files are requested. (I can link an `.har` file if that's better - but I assume the screenshots are much easier to glance over)
If I throttle the page's download speed using the dev tools, I can see the whole page remaining unresponsive while the `.avif`s load, and then as soon as the JavaScript finally loads the page will finally become responsive. When throttle to the "3g" setting, this can take over a minute.

![image](https://github.com/user-attachments/assets/f8d6b310-ea79-415f-af1f-004093587812)
![image](https://github.com/user-attachments/assets/005e10e0-fe51-4ab7-8a4e-741788a05939)
![image](https://github.com/user-attachments/assets/994f909d-9ebe-4236-8bbf-64eb57cfcc9e)
![image](https://github.com/user-attachments/assets/383d6b89-79be-432b-baa2-a7596e7a14af)

Could this be a bug within Edge, or a bug withing `enhanced-img` that only shows itself in Edge?

The reason I assumed this was the fault of `enhanced-img` is because I was initially using plain `<img>` tags in my markup and this problem did not occur. This only started happening after switching to `enhanced-img` - so perhaps it could be that Edge properly handles `<img>` tags, but the `<picture>` tags break it?

**Comment by OOPS-Studio** [comment]
So after messing with this more, it looks like this is what's happening:
In production, the `.avif` requests are being made before the `.js` requests (as can be seen in the screenshots above), _but the JS files are finishing their load first, allowing the site to become responsive while the rest of the images load in._ This is expected behavior and works perfectly.
**However,** in dev mode, the requests are still made in the same order (images -> JavaScript), but for some reason the JavaScript refuses to start loading until after every single image finishes.

And the reason I thought it wasn't working in production either is because production also shows the JavaScript not being requested until after the images - and in dev mode that means it doesn't work. But if I actually watch the page closely in production, I can see that the page does actually become responsive even before the images load, even though the images were requested before the JavaScript.

So I think I was mistaken to assume that just because the JavaScript was requested later, that it also must load later. It looks like it does indeed load first _in production._

But in dev mode, it's still broken. 

This is the best assesment I can come up with right now.

**Comment by Rich-Harris** [comment]
When you import an enhanced image (whether via `src` or an explicit `import` or `import.meta.glob`, you're actually importing a collection of asset URLs. During development, those URLs have to be created on the fly (modulo caching), and since a module cannot be imported/evaluated until its dependencies have been, the result is that the component JavaScript won't evaluate until after the images have been created. This is expected behaviour.

**Comment by OOPS-Studio** [comment]
That makes a lot of sense, thank you Rich. Is there any way to fix that within the `enhandec-img` module in the future, or is it "expected behavior" in the sense that it will never change? If it can be fixed then I'll edit the original issue with this new information so this can be addressed in the future - and if it can't be fixed then we can just close this and call it good.

Currently I can work around it by just testing every now and then by building + hosting locally instead of using the dev server. That's probably not too big of a hassle. Now that I know it's isolated to dev mode I'm not nearly as concerned by it.

**Comment by eltigerchino** [comment]
Yeah, it's unlikely this will change since the images are being optimised as the import is discovered (when loading the page for the first time in dev).

**Comment by benmccann** [comment]
We could have an option to just return the original image in dev mode

**Comment by Rich-Harris** [comment]
I don't think that's a good idea — it'll make it harder to figure out the right configuration. Dev mode should resemble prod mode as closely as possible

**Comment by benmccann** [comment]
Yeah, I probably wouldn't do it by default especially since no one else has complained about it. But maybe if there's some really image heavy page that it's causing trouble with you might need an escape hatch

**Comment by OOPS-Studio** [comment]
Yeah it's a tough situation. Like Rich said - it's better to have dev mode resemble prod as closely as possible. So the question becomes: Which is a closer resemblance? Changing the timing of your JavaScript execution is a pretty big difference imo. I wasted a good 4-6 hours trying to figure out why my JavaScript would not run when I expected it to, and after so much investigating I was pretty sure I had found a bug because it's just very unintuitive the way it is currently.

But perhaps the alternative isn't really much better and just adds confusion in other areas.

I also want to add, just in case it isn't clear, that the delay doesn't necessarily come from the optimization of the image assets as much as it just comes from the _loading_ of the image assets. It doesn't only happen the first time I load the images or after I add an image - it happens every time, even if I just reload the page with no code changes. It's not the 30+ seconds it takes to optimize the images or anything. The duration is determined by what I throttle my bandwidth to in Chrome DevTools, so it's definitely determined strictly by the loading of the images and not the crawling of the app or optimizing or anything. So I don't think it's the image optimizing that's the issue.

But either way, this does seem like an issue that should be fixable. If I haven't changed the file structure of my app at all, I haven't changed which images I'm importing, I haven't changed the content of my images, etc... Then whatever it is that's taking so much time (loading the images, processing the images, optimizing the images - I'm really not sure) should be able to be carried over from the last reload, right? I have no idea how the dev server works or how it differs from a regular server, but this is my first take on the issue as an outsider (and may be completely wrong).

> During development, those URLs have to be created on the fly (modulo caching), and since a module cannot be imported/evaluated until its dependencies have been, the result is that the component JavaScript won't evaluate until after the images have been created.
It seems like this part seems like it should be able to be done once when the images are initially optimized and then carried over between reloads. But maybe I'm wrong.

**Comment by eltigerchino** [comment]
That's unusual. Are your images being optimised again after reloading the page? They should be cached.

**Comment by OOPS-Studio** [comment]
The images aren't optimized again after reloading the page.

Here's a repository where the issue can be seen.
https://github.com/OOPS-Studio/svelte-enhanced-img-issue

If you open Chrome DevTools and throttle the bandwidth to something like "slow 4g" or "3g" then you can see that it takes 10+ seconds before the JavaScript runs. And if you watch the requests as they're made, you can see that the JavaScript files are not even requested at all until after all the .avifs completely finish loading.

Here are 2 videos of me doing it. The first shows how long it takes for "JavaScript loaded" to appear in the console, and the second shows the JavaScript files not being requested until after the .avifs.

Now that I've tried it on a page with far fewer images (previously I was doing it on a page with about 70), I can see that it might not even be a problem with `enhanced-img` specifically and might just be the dev server in general. It seems like the dev server is waiting for absolutely everything to finish loading before it starts loading the JavaScript - and the issue just gets significantly worse when combined with `enhanced-img`. (For example, the delay on this page is about 10 seconds, while the delay on my page with 70 images is like 30 seconds at the same bandwidth limitation).


https://github.com/user-attachments/assets/05730e6d-ae55-46df-9128-8d3980a1507a
https://github.com/user-attachments/assets/69ec7e51-f581-49f6-a41a-26b41763c865

Like Rich said - this is expected behavior. Now the question is just - can it be fixed? Or is this just a necessary evil on the path to all the conveniences SvelteKit provides? As an outsider, I have no clue. All I know is that it hurts the development experience, but it's not the end of the world if it's unfixable.

**Comment by gterras** [comment]
> Here's a repository where the issue can be seen.

I don't reproduce your problem from here. There is an initial delay (probably generating images) then every request is instant, even after modifying project files or when switching browser. So the problem only appears if you artificially lower your bandwidth through devtools after that.

**Comment by OOPS-Studio** [comment]
Indeed, this issue can only be seen when limiting the bandwidth because otherwise the dev server is so fast that your eyes can't see the difference.

But the only reason I noticed this happening in the first place was because I was trying to react to the images loading using JavaScript, and I noticed that my JavaScript was never running. The `load` events on the images were never firing, and after some investigating I realized that it's because the JavaScript wasn't being loaded until after all of the images finished loading. This makes certain tasks impossibel to test in dev mode. For example, I wanted to display loading icons in place of images as they load in, and then remove those icons from the DOM when their respective image loads. Doing this is impossible because of the issue I've described here.

So this gives you a discrepency between dev mode and production. If I put a `console.log()` in a `load` event listener in dev mode, it will never fire. If I put that `console.log()` just straight into my JavaScript to run instantly - it will only run after all the images load. Now if I build my project, suddenly the `console.log()` inside the `load` event listener will fire many times, and the `console.log()` straight in the base of my JavaScript will execute way before any images finish loading. This discrepency is inconvenient and makes testing very difficult - especially when you're building for a non-static host and have no way of testing locally other than the dev server.

I hope I've explained this clearly.
